soh/soh/Enhancements/randomizer/randomizer_check_tracker.cpp

```diff
@@ -0,0 +1,2399 @@
+    { RC_MARKET_BAZAAR_ITEM_4, RR_MARKET },
+    { RC_MARKET_BAZAAR_ITEM_5, RR_MARKET },
+    { RC_MARKET_BAZAAR_ITEM_6, RR_MARKET },
+    { RC_MARKET_BAZAAR_ITEM_7, RR_MARKET },
+    { RC_MARKET_BAZAAR_ITEM_8, RR_MARKET },
+    { RC_KAK_BAZAAR_ITEM_1, RR_KAK },
+    { RC_KAK_BAZAAR_ITEM_2, RR_KAK },
+    { RC_KAK_BAZAAR_ITEM_3, RR_KAK },
+    { RC_KAK_BAZAAR_ITEM_4, RR_KAK },
+    { RC_KAK_BAZAAR_ITEM_5, RR_KAK },
+    { RC_KAK_BAZAAR_ITEM_6, RR_KAK },
+    { RC_KAK_BAZAAR_ITEM_7, RR_KAK },
+    { RC_KAK_BAZAAR_ITEM_8, RR_KAK },
+    { RC_ZD_SHOP_ITEM_1, RR_ZD },
+    { RC_ZD_SHOP_ITEM_2, RR_ZD },
+    { RC_ZD_SHOP_ITEM_3, RR_ZD },
+    { RC_ZD_SHOP_ITEM_4, RR_ZD },
+    { RC_ZD_SHOP_ITEM_5, RR_ZD },
+    { RC_ZD_SHOP_ITEM_6, RR_ZD },
+    { RC_ZD_SHOP_ITEM_7, RR_ZD },
+    { RC_ZD_SHOP_ITEM_8, RR_ZD },
+    { RC_GC_SHOP_ITEM_1, RR_GC },
+    { RC_GC_SHOP_ITEM_2, RR_GC },
+    { RC_GC_SHOP_ITEM_3, RR_GC },
+    { RC_GC_SHOP_ITEM_4, RR_GC },
+    { RC_GC_SHOP_ITEM_5, RR_GC },
+    { RC_GC_SHOP_ITEM_6, RR_GC },
+    { RC_GC_SHOP_ITEM_7, RR_GC },
+    { RC_GC_SHOP_ITEM_8, RR_GC },
+    { RC_COLOSSUS_GOSSIP_STONE, RR_COLOSSUS },
+    { RC_DMC_GOSSIP_STONE, RR_DMC },
+    { RC_DMC_UPPER_GROTTO_GOSSIP_STONE, RR_DMC },
+    { RC_DMT_GOSSIP_STONE, RR_DMT },
+    { RC_DMT_STORMS_GROTTO_GOSSIP_STONE, RR_DMT },
+    { RC_DODONGOS_CAVERN_GOSSIP_STONE, RR_DC },
+    { RC_FAIRY_GOSSIP_STONE, RR_UNKNOWN },
+    { RC_GC_MAZE_GOSSIP_STONE, RR_GC },
+    { RC_GC_MEDIGORON_GOSSIP_STONE, RR_GC },
+    { RC_GV_GOSSIP_STONE, RR_GV },
+    { RC_GY_GOSSIP_STONE, RR_UNKNOWN },
+    { RC_HC_MALON_GOSSIP_STONE, RR_HC },
+    { RC_HC_ROCK_WALL_GOSSIP_STONE, RR_HC },
+    { RC_HC_STORMS_GROTTO_GOSSIP_STONE, RR_HC },
+    { RC_HF_COW_GROTTO_GOSSIP_STONE, RR_HF },
+    { RC_HF_NEAR_MARKET_GOSSIP_STONE, RR_HF },
+    { RC_HF_OPEN_GROTTO_GOSSIP_STONE, RR_HF },
+    { RC_HF_SOUTHEAST_GOSSIP_STONE, RR_HF },
+    { RC_JABU_GOSSIP_STONE, RR_UNKNOWN },
+    { RC_KF_DEKU_TREE_LEFT_GOSSIP_STONE, RR_KF },
+    { RC_KF_DEKU_TREE_RIGHT_GOSSIP_STONE, RR_KF },
+    { RC_KF_GOSSIP_STONE, RR_KF },
+    { RC_KF_STORMS_GOSSIP_STONE, RR_KF },
+    { RC_KAK_OPEN_GROTTO_GOSSIP_STONE, RR_KAK },
+    { RC_LH_LAB_GOSSIP_STONE, RR_LH },
+    { RC_LH_SOUTHEAST_GOSSIP_STONE, RR_LH },
+    { RC_LH_SOUTHWEST_GOSSIP_STONE, RR_LH },
+    { RC_LW_GOSSIP_STONE, RR_LW },
+    { RC_LW_NEAR_SHORTCUTS_GOSSIP_STONE, RR_LW },
+    { RC_SFM_MAZE_LOWER_GOSSIP_STONE, RR_SFM },
+    { RC_SFM_MAZE_UPPER_GOSSIP_STONE, RR_SFM },
+    { RC_SFM_SARIA_GOSSIP_STONE, RR_SFM },
+    { RC_TOT_LEFT_CENTER_GOSSIP_STONE, RR_TOT },
+    { RC_TOT_LEFT_GOSSIP_STONE, RR_TOT },
+    { RC_TOT_RIGHT_CENTER_GOSSIP_STONE, RR_TOT },
+    { RC_TOT_RIGHT_GOSSIP_STONE, RR_TOT },
+    { RC_ZD_GOSSIP_STONE, RR_ZD },
+    { RC_ZR_NEAR_DOMAIN_GOSSIP_STONE, RR_ZR },
+    { RC_ZR_NEAR_GROTTOS_GOSSIP_STONE, RR_ZR },
+    { RC_ZR_OPEN_GROTTO_GOSSIP_STONE, RR_ZR }
+};
+
+std::unordered_map<RandomizerCheck, std::string> CheckEnumToName = {
```

Comment:
i think instead of using this it might make sense to move `SpoilerfileCheckNameToEnum` from `randomizer.cpp` https://github.com/HarbourMasters/Shipwright/blob/e4b58e5a0ce29eeed4e3fb82dbb02961758098b0/soh/soh/Enhancements/randomizer/randomizer.cpp#L71 to somewhere it can be accessed by both everything in `randomizer.cpp` and here

Reply:
come to think of it, it probably makes sense to just make a public `GetCheckName` function in `randomizer.cpp` that takes in a `RandomizerCheck` and gets the string from the existing map

Reply:
Would it make more sense to go in randomizer.cpp or maybe randomizerItemTracker.cpp? Or maybe even a completely new file to start the process of breaking randomizer.cpp up a bit? The latter might be better saved for later, relocation is probably relatively easy.

Reply:
i was thinking going into `randomizer.cpp` makes sense because that's where `SpoilerfileCheckNameToEnum` is. and yeah, splitting up the behemoth that is `randomizer.cpp` is definitely something we should do, but it can be it's own thing

Reply:
Surely we can find a way to re-use the already existing list of checks and locations from within 3drano right? A lot of this has been done just with a different purpose, can we repurpose it?

Reply:
There's a few things that aren't ideal about that. Where we have separate RC and RG Enums, they have one giant Keys enum that has item locations and the items themselves, and I think entrances as well? Then, for each of those things they have separate arrays for Items, Locations, etc. that all have the size of the largest Keys Enum value, meaning that, unless I'm mistaken, a significant chunk of each of those arrays is unused. This isn't as big a deal for them as it would be for us, since those arrays only exist while generating the seed I believe, not in-game. We actually have to store these arrays in the SaveContext and reference them by the enum values during runtime, so we need to use more tightly-sized arrays. Hence the separate enums for each type of Key. We technically could've still done this while using the existing Keys enum, but then we would've needed to do some magic-number-y math in a few places, which would make things less clear.

Reply:
Anyway, back to the review at hand. SpoilerfileCheckNameToEnum maps in the opposite direction from this enum if I'm not mistaken. I'm looking at cppreference.com and I'm not seeing any way to get the key (location name) from just the value (check) so I don't think they can direclty use SpoilerfileCheckNameToEnum. They could probably loop over it and construct a new map setting the old value to the new key and the old key to the new value rather than writing it by hand. They could also make it an array, since the keys in this case are just numbers under the hood, and the size will be known at compile time.

Reply:
Actually I suppose you could just have the `GetCheckName` function you suggested iterate over the `SpoilerfileCheckNameToEnum` map until it finds a specific value, but that seems a bit inefficient considering how often it will be done as compared to my suggestion of iterating over the map once on startup and constructing a new array. What do you think @briaguya-ai ?

Reply:
yeah, not sure exactly how bad using [find](https://cplusplus.com/reference/unordered_map/unordered_map/find/) would be from a performance stand point in that scenario, it's not like it'd be doing string comparisons so it might not be *too* bad.

but the map isn't big enough to be a memory problem on the systems we're targeting, so generating a new array (or a new unordered_map with the keys/values reversed) based on the map in the randomizer constructor would probably be a good plan

Reply:
`find` wouldn't work would it? `find` matches on the key, if they're going to use our existing map they would be searching for a specific *value* right?

Array would be best I think since the keys in this case are just integers and nothing will be getting added after compile time.

Reply:
> find matches on the key

yeah, my bad, good point

> Array would be best I think

yeah, array sounds good, i'd still like to expose it via a get method in randomizer.cpp but storing it as an array seems like a solid plan

Reply:
I concur on the Get Method, that's still a solid plan.